simulators/ethereum/engine/clmock.go

```diff
@@ -0,0 +1,351 @@
+package main
+
+import (
+	"fmt"
+	"math/big"
+	"math/rand"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+)
+
+// Consensus Layer Client Mock used to sync the Execution Clients once the TTD has been reached
+type CLMocker struct {
+	EngineClients           []*EngineClient
+	RandomHistory           map[uint64]common.Hash
+	BlockProductionMustStop bool
+
+	// Latest broadcasted data using the PoS Engine API
+	LatestFinalizedNumber *big.Int
+	LatestFinalizedHeader *types.Header
+	LatestExecutedPayload catalyst.ExecutableDataV1
+	LatestForkchoice      catalyst.ForkchoiceStateV1
+
+	// Merge related
+	FirstPoSBlockNumber         *big.Int
+	TTDReached                  bool
+	PoSBlockProductionActivated bool
+	NextFeeRecipient            common.Address
+
+	// PoS Status Mutexes: Set-Reset lock is used to guarantee each test finds the environment as expected,
+	// and not as previously modified by another test.
+	PayloadBuildMutex                SetResetLock
```

Comment:
Not sure I understand the SetResetLock concept. What are the requirements on this mutex? What semantics do you want to achieve?

Reply:
This is an attempt to provide a clean slate for all test cases to start from.

The test cases request a lock to "wake up" at a specific point during the PoS block production procedure and, when the CL mocker has reached that point, it releases the lock to let the test cases' procedures do their testing.

The problem that came up when using a regular lock was that most of the test cases send API calls to test some functionality, and in doing so, the state at which the next test case finds the client is different from what the CL mocker left, and from what the test case expected. This resulted in tests failing.

The idea behind the set-reset is that only one test case runs each time the CL mocker reaches a certain point in its block production.

The CL mocker 'resets' the lock to signal that nobody else has gotten the lock since the last time it was released, and therefore the state of the client is exactly as expected.

If the test case gets the lock but it is already 'set', it knows that another test case got to run before them, and it can pass on getting the lock this time and try again until another block is produced and it gets the lock in a 'reset' state.